### PR TITLE
tests: add tests for reliable service under packet loss

### PR DESCRIPTION
**Pull request by ainghazal**
# Checklist

* [x] I have read the contribution guidelines
* [x] Iff you changed code related to services, or inter-service communication, make sure you update the diagrams in `ARCHITECTURE.md`.
* [x] Reference issue for this pull request:

# Description

Exercise reliable service passing a vector of packet ID to simulate packet loss.

Additionally, this PR updates github actions version to the latest action release for checkout (v4) and setup-go (v5). Eventually, we'll currently replace the test setup with the one we're using to run tests on the internal packages. 

TODO: This PR needs to be rebased on top of #60

### CODE REVIEW

**Comment by ainghazal** [comment]
I think this is ready for review. the only file of interest after merging #60 should be https://github.com/ooni/minivpn/pull/63/files#diff-b9a3711895939a4f4dc918dccb1b8e3eda12d2fea35a79984d7a7a7aab12a9a3 (modulo any minor changes)

File: internal/vpntest/packetio.go [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+	pw.ch <- p
+	time.Sleep(testPkt.IAT)
+}
+
+// WriteSequenceWithFixed payload will write packets according to the sequence specified in seq,
+// but will sequentially pick the payload from the passed payload string, in increments defined by size.
+func (pw *PacketWriter) WriteSequenceWithFixedPayload(seq []string, payload string, size int) {
+	pw.payload = payload
+	pw.packetPayloadSize = 3
+	pw.WriteSequence(seq)
```
Comment (ainghazal) [inline internal/vpntest/packetio.go:117]:
```suggestion
	pw.packetPayloadSize = size
	pw.WriteSequence(seq)
```

File: internal/reliabletransport/reliable_loss_test.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+					"[9] CONTROL_V1 +10ms",
+					"[10] CONTROL_V1 +10ms",
+				},
+				inputPayload: "fffcccaaabbbdddeeeggghhhiiijjj",
+				want:         "aaabbbcccdddeeefffggghhhiiijjj",
+				losses:       []int{1, 5},
+			},
+		},
+
+		// we could exclude the following tests if not --short
```
Comment (bassosimone) [inline internal/reliabletransport/reliable_loss_test.go:185]:
```suggestion
		// TODO(ainghazal): exclude the following tests if not `-short`?
```

File: internal/reliabletransport/reliable_reorder_test.go [hunk 3]
```diff
@@ -10,9 +10,20 @@ import (
...
+//
+//		           dataOut
+//		              ▲
+//		              │
+//		 ┌────┐    ┌──┴─┐
+//		 │sndr│    │rcvr│
+//		 └────┘    └────┘
+//	              ▲
+//	             |
+//		          dataIn
```
Comment (bassosimone) [inline internal/reliabletransport/reliable_reorder_test.go:22]: something's wrong in this diagram

File: internal/vpntest/packetio.go [hunk 4]
```diff
@@ -112,27 +201,45 @@ func (pr *PacketReader) WaitForSequence(seq []int, start time.Time) bool {
 			break
 		}
 		// no, so let's keep reading until the test runner kills us
-		pkt := <-pr.ch
-		pr.log = append(pr.log, newLoggedPacket(pkt, start))
-		log.Debugf("got packet: %v", pkt.ID)
+		pr.appendOneIncomingPacket(start)
 	}
-	// TODO(ainghazal): move the comparison to witness, leave only wait here
+	// TODO move the comparison to witness, leave only wait here
```
Comment (bassosimone) [inline internal/vpntest/packetio.go:206]:
```suggestion
	// TODO(ainghazal): move the comparison to witness, leave only wait here
```

File: internal/vpntest/packetio_test.go [hunk 5]
```diff
@@ -79,3 +126,333 @@ func TestPacketReaderWriter(t *testing.T) {
...
+			},
+			args: args{
+				packetsIn: []int{1, 2, 3, 2, 1, 4, 5, 1, 2},
+				losses:    []int{1, 1, 2, 2},
+				wantOut:   []int{3, 4, 5, 1, 2},
+			},
+		},
+	}
+	for _, tt := range tests {
+
```
Comment (bassosimone) [inline internal/vpntest/packetio_test.go:315]:
```suggestion
```

**Review by bassosimone — APPROVED** [review]
🎻 🎻 🎻 🎻 

LGTM with the provided comments applied

File: internal/reliabletransport/reliable_reorder_test.go (hunk 3 above)
Comment (ainghazal) [reply internal/reliabletransport/reliable_reorder_test.go]: your dear tabs :P fixed

**Pull request merged**